File: solutions/42/linear_time_constant_space.js
let trap = function(height) {
    let n = height.length;
    let max = Math.max(...height);
    let mi = height.indexOf(max);
    let count = 0;
    let lmax = 0;
    for(let i=0; i<mi; i++){
        if(height[i] > lmax){
            lmax =  height[i];
        }else{
            count += lmax - height[i]
        }
    }
    let rmax = 0;
    for(let j=n-1; j>mi; j--){
        if(height[j] > rmax){
            rmax = height[j];
        }else{
            count +=  rmax - height[j];
        }
    }
    return count;
}

File: solutions/42/linear_time_constant_space.py
def f(height):
  n = len(height)

  # index of last max element
  max_ix = max(range(n - 1, -1, -1),
               key=lambda e: height[e])

  water = 0
  lmax = 0
  for i in range(max_ix):
    if height[i] > lmax:
      lmax = height[i]
      continue
    water += lmax - height[i]

  rmax = 0
  for i in range(n - 1, max_ix, -1):
    if height[i] > rmax:
      rmax = height[i]
      continue
    water += rmax - height[i]

  return water

assert f((0,1,0,2,1,0,1,3,2,1,2,1)) == 6
assert f((4,2,0,3,2,5)) == 9
